src/devices/Ak8963/Ak8963.cs

```diff
@@ -0,0 +1,287 @@
+            calib.Z = (float)((rawData[2] - 128) / 256.0 + 1.0);
+            MeasurementMode = MeasurementMode.PowerDown;
+            MeasurementMode = oldPower;
+
+            // Now calculate the bias
+            // Store old mode to restore after
+            var oldMode = MeasurementMode;
+            Vector3 minbias = new Vector3();
+            Vector3 maxbias = new Vector3();
+            int numberMeasurements = 1500;
+
+            // Setup the 100Hz continuous mode
+            MeasurementMode = MeasurementMode.ContinuousMeasurement100Hz;
+            for (int reading = 0; reading < numberMeasurements; reading++)
+            {
+                // Timeout = 100Hz = 10 ms + 1 for security
+                var bias = ReadMagnetometer(true, TimeSpan.FromMilliseconds(20));
+                minbias.X = Math.Min(bias.X, minbias.X);
+                minbias.Y = Math.Min(bias.Y, minbias.Y);
+                minbias.Z = Math.Min(bias.Z, minbias.Z);
+                maxbias.X = Math.Max(bias.X, maxbias.X);
+                maxbias.Y = Math.Max(bias.Y, maxbias.Y);
+                maxbias.Z = Math.Max(bias.Z, maxbias.Z);
+                // 10 ms = 100Hz, so waiting to make sure we have new data
+                Thread.Sleep(10);
+            }
+            // Store the bias
+            MagnometerBias = ((maxbias + minbias) / 2) + calib;
+
+            return calib;
+        }
+
+        /// <summary>
+        /// True if there is a data to read
+        /// </summary>
+        public bool HasDataToRead => (ReadByte(Register.ST1) & 0x01) == 0x01;
+
+        /// <summary>
+        /// Check if the version is the correct one (0x48). This is fixed for this device
+        /// Page 28 from the documentation :
+        /// Device ID of AKM. It is described in one byte and fixed value.  48H: fixed 
+        /// </summary>
+        /// <returns>Returns true if the version match</returns>
+        public bool CheckVersion()
+        {
+            return ReadByte(Register.WIA) == 0x48;
+        }
+
+        /// <summary>
+        /// Read the magnetometer and can wait for new data to be present
+        /// </summary>
+        /// <remarks>
+        /// Vector axes are the following:
+        ///         +X
+        ///  \  |  /
+        ///   \ | /
+        ///    \|/
+        ///    /|\
+        ///   / | \
+        ///  /  |  \
+        ///    +Z   +Y
+        /// </remarks>
+        /// <param name="waitForData">true to wait for new data</param>
+        /// <param name="timeout">timeout for waiting the data, ignored if waitForData is false</param>
+        /// <returns>The data from the magnetometer</returns>
+        public Vector3 ReadMagnetometer(bool waitForData, TimeSpan timeout)
+        {
+            Span<byte> rawData = stackalloc byte[6];
+            // Wait for a data to be present
+            if (waitForData)
+            {
+                DateTime dt = DateTime.Now.Add(timeout);
+                while (!HasDataToRead)
```

Comment:
I'd recommend to add SpinWait in the loop so the CPU doesn't go 100%

Reply:
I've changed it in the last commit, so not sure why it's not there:
```csharp
            Span<byte> rawData = stackalloc byte[6];
            // Wait for a data to be present
            if (waitForData)
            {
                DateTime dt = DateTime.Now.Add(timeout);
                while (!HasDataToRead)
                {
                    if (DateTime.Now > dt)
                        throw new Exception($"{nameof(ReadMagnetometer)} timeout reading value");
                }
            }
```